### PR TITLE
Add Container and Block for Text

### PR DESCRIPTION
**Pull request by Chandu-4444**
Tried starting at creating a simple textual recipe based on ImageFolders dataset recipe. This specifically works for `imdb` and similar datasets. Any feedback is highly appreciated.

### CODE REVIEW

**Comment by Chandu-4444** [comment]
```julia
julia> using FastAI

julia> name, recipe = finddatasets(blocks=(Any, Any), name="imdb")[1]
Pair{String, FastAI.Datasets.DatasetRecipe}("imdb", TextFolders(FastAI.Datasets.parentname, false, FastAI.Text.var"#2#4"()))

julia> data, blocks = loadrecipe(recipe, datasetpath("imdb"))
((mapobs(loadfile, ["/home/luna/.julia/datadeps/fastai-imdb/imdb/test/neg/0_2.txt", "/home/luna/.ju…]), mapobs(parentname, ["/home/luna/.julia/datadeps/fastai-imdb/imdb/test/neg/0_2.txt", "/home/luna/.ju…])), (TextBlock(), Label{String}(["neg", "pos"])))

julia> text, class = obs = getobs(data, 1000)
("Every movie I have PPV'd because Leonard Maltin praised it to the skies has blown chunks! 
Every single one! 
When will I ever learn?<br /><br />Evie is a raving Old Bag who thinks nothing of saying she's dying of breast cancer to get her way! 
Laura is an insufferable Medusa filled with  The Holy Spirit (and her hubby's protégé)! 
Caught between these harpies is Medusa's dumb-as-a-rock boy who has been pressed into weed-pulling servitude by The Old Bag!<br /><br />
As I said, when will I ever learn?<br /><br />
I was temporarily lifted out of my malaise when The Old Bag stuck her head in a sink, but, unfortunately, she did not die. 
I was temporarily lifted out of my malaise again when Medusa got mowed down, but, unfortunately, she did not die. 
It should be a capital offense to torture audiences like this!<br /><br />
Without Harry Potter to kick him around, Rupert Grint is just a pair of big blue eyes that practically bulge out of its sockets.  
Julie Walters's scenery-chewing (especially the scene when she \"plays\" God) is even more shameless than her character.
<br /><br />
At least this Harold bangs some bimbo instead of Maude. 
For that, I am truly grateful. And if you're reading this Mr. Maltin, you owe me \$3.99!", "neg")


```

**Comment by Chandu-4444** [comment]
I started adding various text transforms as in fastai text module.

**Comment by Chandu-4444** [comment]
I have started adding functions for replacing words that starts with uppercase letters, contains all uppercase letters with special tokens like xxup, xxmaj etc.

**Comment by Chandu-4444** [comment]
Each observation now has xxbos special token to indicate the start of the sentence just like fastai do.

File: src/Text/blocks/text.jl [hunk 1]
```diff
@@ -0,0 +1 @@
+struct TextBlock <: Block end
```
Comment (lorenzoh) [inline src/Text/blocks/text.jl:1]: Some more documentation on what this represents would be great. Is this a sentence, a paragraph? A `checkblock` is also needed. The convention is not to put `Block` in the name of `Block`s, so depending on what this represents, we may be able to find a more precise term.

What tasks does this adress? Is there a need for different text-related blocks to solve other tasks? I'm. not super familiar with deep NLP, but these questions may help making this more precise.

File: src/Text/Text.jl [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+include("transform.jl")
+
+function __init__()
+    _registerrecipes()
+    @require Makie="ee78f7c6-11fb-53f2-987a-cfe4a2b5a57a" begin
+        import .Makie
+        import .Makie: @recipe, @lift
+        import .FastAI: ShowMakie
+        include("makie.jl")
+    end
```
Comment (lorenzoh) [inline src/Text/Text.jl:29]: We can drop this until we need to add Makie.jl recipes

File: src/Text/recipes.jl [hunk 3]
```diff
@@ -0,0 +1,45 @@
+"""
+    TextFolders(textfile; labelfn = parentname, split = false)
+
+Recipe for loading a single-label text classification dataset
+stored in hierarchical folder format. 
+"""
+
```
Comment (lorenzoh) [inline src/Text/recipes.jl:7]:
```suggestion
```

File: src/Text/Text.jl [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+end
+
+export TextBlock, TextFolders, replace_all_caps, replace_sentence_case,
+    convert_lowercase
```
Comment (lorenzoh) [inline src/Text/Text.jl:33]: We should hold off on exporting everything except the block
```suggestion
export TextBlock
```

**Review by lorenzoh — CHANGES_REQUESTED** [review]
Thanks for the PR!

I've left some comments.

Is there a fastai tutorial that uses this dataset? Would be helpful to know what kind of tasks could be tackled with this.

**Comment by Chandu-4444** [comment]
> Is there a fastai tutorial that uses this dataset? Would be helpful to know what kind of tasks could be tackled with this.

Yes, fastai does have a tutorial that uses this dataset (And it's the first tutorial that focuses on the text module of the fastai library), https://docs.fast.ai/tutorial.text.html. This tutorial focuses on the sentiment analysis based on the given review. This first part uses a pre-trained language model (called [AWD-LSTM](https://arxiv.org/abs/1708.02182)) on Wikipedia for predicting the next word (language generation), and is directly used for predicting the sentiment for the given review. In the second part of the tutorial, they used a [ULMFit](https://arxiv.org/abs/1801.06146) approach that involves fine-tuning the model with the IMDB dataset and using that for predicting the sentiment. They achieved SOTA using the second method. 

I'll commit to the suggestions provided above and will improve this.  

Simultaneously, I'll start looking into that AWD-LSTM (https://arxiv.org/abs/1708.02182) paper to get deeper into how the model works. After that, the plan was to go through the ULMFit (https://arxiv.org/abs/1801.06146) paper.

File: src/Text/Text.jl (hunk 2 above)
Comment (Chandu-4444) [reply src/Text/Text.jl:29]: Ok. I'll remove this.

File: src/Text/blocks/text.jl (hunk 1 above)
Comment (Chandu-4444) [reply src/Text/blocks/text.jl:1]: > Some more documentation on what this represents would be great. Is this a sentence, a paragraph?

The initial idea was to replicate `struct Image{N} <: Block end`. I haven't specified the dimension `N` as I'm not sure how to incorporate it with textual data is it the length? Can that be the vocabulary size? Or can we just leave this as it is? (Maybe we can discuss this). 

> Is there a need for different text-related blocks to solve other tasks?

Most text-related tasks revolve around classification and sentiment analysis (Which are single label prediction tasks). There are some more tasks that involve predicting multiple labels like topic modeling (Need more than one label to be predicted). Finally, there are some unsupervised techniques like text generation, text summarization.

I guess all the observations for the above tasks can be represented using a block of the same type?

> Is this a sentence, a paragraph?

I think the better name would be `Paragraph` rather than `TextBlock` as most if not all of the text data has a single paragraph (multiple sentences) as an individual observation (We can consider a sentence as a paragraph too!).

Comment (Chandu-4444) [reply src/Text/blocks/text.jl]: >  A checkblock is also needed.

I tried adding the `checkblock` for `TextBlock`, but it is always returning false. I guess it is calling the default `checkblock` method. I'm sure why this happens. Also the method I wrote isn't even getting displayed in the output of `methods(checkblock)`.

File: src/Text/recipes.jl [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+function Datasets.loadrecipe(recipe::TextFolders, path)
+    isdir(path) || error("$path is not a directory")
+    data = loadfolderdata(
+        path,
+        filterfn=f -> istextfile(f) && recipe.filefilterfn(f),
+        loadfn=(loadfile, recipe.labelfn),
+        splitfn=recipe.split ? grandparentname : nothing)
+
+    (recipe.split ? length(data) > 0 : nobs(data) > 0) || error("No text files found in $path")
```
Comment (ToucheSir) [inline src/Text/recipes.jl:23]: Doesn't `nobs` fall back to `length` now (or vice versa)? Trying to understand why the nested ternary is necessary.

File: src/Text/recipes.jl [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    labels = recipe.split ? first(values(data))[2] : data[2]
+    blocks = TextBlock(), Label(unique(eachobs(labels)))
```
Comment (ToucheSir) [inline src/Text/recipes.jl:26]: Was this meant to be 
```suggestion
    blocks = (TextBlock(), Label(unique(eachobs(labels))))
```
? Given the lines below, I'd recommend declaring them separately and doing the tuple packing in the return.

File: src/Text/recipes.jl [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    length(blocks[2].classes) > 1 || error("Expected multiple different labels, got: $(blocks[2].classes))")
+    return data, blocks
+end
+
+# Registering recipes
+
+const RECIPES = Dict{String,Vector{Datasets.DatasetRecipe}}(
+    "imdb" => [TextFolders(
+        filefilterfn = f->!contains(f, "tmp_clas") && !contains(f, "tmp_lm") && !contains(f, "unsup")
```
Comment (ToucheSir) [inline src/Text/recipes.jl:35]:
```suggestion
        filefilterfn = f -> !occursin(f, "tmp_clas|tmp_lm|unsup")
```
A comment about why directories with these file names are to be avoided would be good too.

File: src/Text/transform.jl [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+"""
+
+function replace_sentence_case(t::String)
+    t = replace(t, r"(?<!\w)([A-Z][A-Z0-9]*[a-z0-9]+)(?!\w)" => s"xxmaj \1")
+    t = replace(t, r"(?<!\w)([A-Z][A-Z0-9]*[a-z0-9]+)(?!\w)" => lowercase)
+end
+
+function convert_lowercase(t::String)
+    string("xxbos ", lowercase(t))
+end
```
Comment (ToucheSir) [inline src/Text/transform.jl:25]:
```suggestion
function replace_all_caps(t::String)
    t = replace(t, r"([A-Z]+[^a-z\s]*)(?=(\s|$))" => s"xxup \1")
    return replace(t, r"([A-Z]*[^a-z\s]+)(?=(\s|$))" => lowercase)
end

"""
    replace_sentence_case(String)

Replace tokens in Sentence Case by their lower verions and add xxmaj before.
"""
function replace_sentence_case(t::String)
    t = replace(t, r"(?<!\w)([A-Z][A-Z0-9]*[a-z0-9]+)(?!\w)" => s"xxmaj \1")
    return replace(t, r"(?<!\w)([A-Z][A-Z0-9]*[a-z0-9]+)(?!\w)" => lowercase)
end

convert_lowercase(t::String) = string("xxbos ", lowercase(t))

```
Not sure if the type annotations are necessary.

File: src/datasets/containers.jl [hunk 9]
```diff
@@ -38,6 +38,8 @@ function loadfile(file::String)
         return FileIO.load(file, view = true)
     elseif endswith(file, ".csv")
         return DataFrame(CSV.File(file))
+    elseif endswith(file, ".txt")
+        return String(read(file))
```
Comment (ToucheSir) [inline src/datasets/containers.jl:42]:
```suggestion
        return read(file, String)
```
Per last signature in https://docs.julialang.org/en/v1/base/io-network/#Base.read.

File: src/Text/recipes.jl (hunk 7 above)
Comment (Chandu-4444) [reply src/Text/recipes.jl]: Including data from `unsup` folder for unsupervised text classification can be done in the later stages. And the other two directories were just some metadata.

File: src/Text/recipes.jl (hunk 5 above)
Comment (Chandu-4444) [reply src/Text/recipes.jl:23]: Here, If the split is set to true then it checks for the corresponding loadfile mapobs and parentname mapobs by using `length(data)` which returns 2 when we have a valid `data` when using split. The next case would be checking for total obs in the data, done by `nobs(data)` return 50000 in this case. If none of this happens, we throw an error.

This is how I've seen it was implemented in `Vision/recipes.jl`.

File: src/Text/recipes.jl (hunk 6 above)
Comment (Chandu-4444) [reply src/Text/recipes.jl:26]: Even this is how I've seen it was implemented in `Vision/recipes.jl`. But if it seems reasonable to do this way, we can change it.

File: src/Text/recipes.jl (hunk 5 above)
Comment (ToucheSir) [reply src/Text/recipes.jl:23]: Got it. I think there needs to be a bit more documentation on this across recipes (or at least more well-named things in code so that one can figure out what all the `[2]` business is about, but that need not concern this PR.

Comment (lorenzoh) [reply src/Text/recipes.jl:23]: FastAI.jl does not use MLUtils.jl yet, see #196. Yes for the docs; in short, `blocks` is a tuple of `(inputblock, targetblock)` so `blocks[2]` takes the target block.

File: src/Text/blocks/text.jl (hunk 1 above)
Comment (lorenzoh) [reply src/Text/blocks/text.jl]: > The initial idea was to replicate struct Image{N} <: Block end. I haven't specified the dimension N as I'm not sure on incorporating it with textual data. Is it the length? Can that be the vocabulary size? Or can we just leave this as it is? (Maybe we can discuss this).

Don't think this is necessary. As described in the documentation of `Block`, the block should only hold information that is constant for all observations, so encoding vocab size here would mean we assume it is fixed for all observations in a dataset which is too restrictive.

> I guess all the observations for the above tasks can be represented using a block of the same type?

Yeah, that sounds like it 👍 

> I think the better name would be Paragraph rather than TextBlock as most if not all of the text data has a single paragraph (multiple sentences) as an individual observation (We can consider a sentence as a paragraph too!).

`Paragraph` is good :)

**Review by lorenzoh — CHANGES_REQUESTED** [review]
Left a comment regarding the block type, but other than that and Brian's open suggestion, this looks good to me!

File: src/Text/blocks/text.jl (hunk 1 above)
Comment (Chandu-4444) [reply src/Text/blocks/text.jl]: > Paragraph is good :)

Yes, that's what I thought would be more reasonable. Changed it now.

**Comment by lorenzoh** [comment]
Sorry for letting this sit!

Tests were failing due to issues that should be fixed on master, so merging master into this should make the CI green.

Last thing that would be good to have would be some tests

**Comment by Chandu-4444** [comment]
Sure! Will synchronise it with master and add some tests.

**Comment by Chandu-4444** [comment]
Umm... For writing tests to the TextFolders(), I need to access the IMDb dataset. I remember Lorenz mentioning that it isn't very nice to use large datasets for testing as it might overload the CI system. And for other recipes, there are smaller version datasets that replicate the original larger version datasets. I couldn't find any such datasets for IMDb (Actually there is one such dataset that is available as a CSV file, but I need an IMDb-like directory structure for testing the recipe). Is there any workaround?

**Comment by ToucheSir** [comment]
I wouldn't worry about testing the bits that require file IO for now, mostly the helper functionality.

**Comment by Chandu-4444** [comment]
That sounds good!

File: src/Textual/blocks/text.jl [hunk 10]
```diff
@@ -26,4 +26,4 @@ FastAI.mockblock(Paragraph())
 struct Paragraph <: Block end
 
 FastAI.checkblock(::Paragraph, ::String) = true
-FastAI.mockblock(::Paragraph) = randstring(" ABCEEFGHIJKLMNOPQESRUVWXYZ 1234567890abcdefghijklmnopqrstynwxyz\n\t.,", 40)
+FastAI.mockblock(::Paragraph) = randstring(" ABCEEFGHIJKLMNOPQESRUVWXYZ 1234567890 abcdefghijklmnopqrstynwxyz\n\t.,", 40)
```
Comment (lorenzoh) [inline src/Textual/blocks/text.jl:29]: `Paragraph` can have variable length, right? Let's put this in `mockblock` so that we don't bake in invariants that don't need to hold: 
```suggestion
FastAI.mockblock(::Paragraph) = randstring(" ABCEEFGHIJKLMNOPQESRUVWXYZ 1234567890 abcdefghijklmnopqrstynwxyz\n\t.,", rand(10:40))
```

File: src/Textual/recipes.jl [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        Datasets.registerrecipe!(Datasets.FASTAI_DATA_REGISTRY, name, recipe)
+    end
+end
+
+
+## Tests
+
+
+@testset "TextFolders [Recipe]" begin
+    @test length(Datasets.FASTAI_DATA_REGISTRY.recipes["imdb"]) == 1
```
Comment (lorenzoh) [inline src/Textual/recipes.jl:50]: This is brittle in that it will break when a new recipe is added to `"imdb"` and goes into the internals of the dataset registry. This should test `finddatasets("imdb", B)` returns _at least one_ recipe for some block `B` we want to test.

File: src/Textual/transform.jl [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+
+Replace tokens in ALL CAPS by their lower version and add xxup before.
+"""
+
+function replace_all_caps(t)
+    t = replace(t, r"([A-Z]+[^a-z\s]*)(?=(\s|$))" => s"xxup \1")
+    return replace(t, r"([A-Z]*[^a-z\s]+)(?=(\s|$))" => lowercase)
+end
+
+"""
```
Comment (lorenzoh) [inline src/Textual/transform.jl:12]: Some inline tests for these preprocessing steps would be good to have, too

**Review by lorenzoh — CHANGES_REQUESTED** [review]

File: src/Textual/blocks/text.jl (hunk 10 above)
Comment (Chandu-4444) [reply src/Textual/blocks/text.jl:29]: Yeah, that's more reasonable thing to do.

File: src/Textual/transform.jl (hunk 12 above)
Comment (Chandu-4444) [reply src/Textual/transform.jl:12]: Yeah, that's in my to-do list. Just wanted to make sure they're necessary. I'll add some tests in my next commit.

File: src/Textual/recipes.jl (hunk 11 above)
Comment (Chandu-4444) [reply src/Textual/recipes.jl:50]: Sure. I'll update the test.
